File: iosApp/iosApp/Modules/NFTsDashboard/NFTsDashboardWireframe.swift
// Created by web3d4v on 24/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultNFTsDashboardWireframe {
    private weak var parent: UIViewController?
    private let nftsCollectionWireframeFactory: NFTsCollectionWireframeFactory
    private let nftDetailWireframeFactory: NFTDetailWireframeFactory
    private let nftsService: NFTsService
    private let networksService: NetworksService
    private let mailService: MailService

    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        nftsCollectionWireframeFactory: NFTsCollectionWireframeFactory,
        nftDetailWireframeFactory: NFTDetailWireframeFactory,
        nftsService: NFTsService,
        networksService: NetworksService,
        mailService: MailService
    ) {
        self.parent = parent
        self.nftsCollectionWireframeFactory = nftsCollectionWireframeFactory
        self.nftDetailWireframeFactory = nftDetailWireframeFactory
        self.nftsService = nftsService
        self.networksService = networksService
        self.mailService = mailService
    }
}

extension DefaultNFTsDashboardWireframe {

    func present() {
        let vc = wireUp()
        if let tabVc = parent as? UITabBarController {
            let vcs = (tabVc.viewControllers ?? []) + [vc]
            tabVc.setViewControllers(vcs, animated: false)
        } else {
            parent?.show(vc, sender: self)
        }
    }

    func navigate(with destination: NFTsDashboardWireframeDestination) {
        if let input = destination as? NFTsDashboardWireframeDestination.ViewNFT {
            let context = NFTDetailWireframeContext(
                nftId: input.nftItem.identifier,
                collectionId: input.nftItem.collectionIdentifier
            )
            nftDetailWireframeFactory.make(vc, context: context).present()
        }
        if let input = destination as? NFTsDashboardWireframeDestination.ViewCollectionNFTs {
            let context = NFTsCollectionWireframeContext(collectionId: input.collectionId)
            nftsCollectionWireframeFactory.make(vc, context: context).present()
        }
        if let input = destination as? NFTsDashboardWireframeDestination.SendError {
            mailService.sendMail(context: .init(subject: .app, body: input.msg))
        }
    }
}

private extension DefaultNFTsDashboardWireframe {

    func wireUp() -> UIViewController {
        let vc: NFTsDashboardViewController = UIStoryboard(
            .nftsDashboard
        ).instantiate()
        let interactor = DefaultNFTsDashboardInteractor(
            networksService: networksService,
            nftsService: nftsService
        )
        let presenter = DefaultNFTsDashboardPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        nc.tabBarItem = UITabBarItem(
            title: Localized("nfts"),
            image: "tab_icon_nfts".assetImage,
            tag: 2
        )
        return nc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftsdashboard/NFTsDashboardWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftsdashboard

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail.NFTDetailWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftscollection.DefaultNFTsCollectionWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeContext
import com.sonsofcrypto.web3walletcore.modules.nftsCollection.NFTsCollectionWireframeContext
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.DefaultNFTsDashboardInteractor
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.DefaultNFTsDashboardPresenter
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframe
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframeDestination
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

class DefaultNFTsDashboardWireframe(
    private val parent: WeakRef<Fragment>?,
    private val networksService: NetworksService,
    private val nftsService: NFTsService,
    private val nftsCollectionWireframeFactory: DefaultNFTsCollectionWireframeFactory,
    private val nftDetailWireframeFactory: NFTDetailWireframeFactory
): NFTsDashboardWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment)
    }

    override fun navigate(destination: NFTsDashboardWireframeDestination) {
        when (destination) {
            is NFTsDashboardWireframeDestination.ViewCollectionNFTs -> {
                nftsCollectionWireframeFactory.make(
                    parent?.get(),
                    NFTsCollectionWireframeContext(destination.collectionId)
                ).present()
            }
            is NFTsDashboardWireframeDestination.ViewNFT -> {
                val context = NFTDetailWireframeContext(
                    destination.nftItem.identifier, destination.nftItem.collectionIdentifier
                )
                nftDetailWireframeFactory.make(parent?.get(), context).present()
            }
            is NFTsDashboardWireframeDestination.SendError -> {

            }
        }
    }

    private fun wireUp(): Fragment {
        val view = NFTsDashboardFragment()
        val interactor = DefaultNFTsDashboardInteractor(
            networksService,
            nftsService,
        )
        val presenter = DefaultNFTsDashboardPresenter(
            WeakRef(view),
            this,
            interactor,
        )
        view.presenter = presenter
        return NavigationFragment(view)
    }
}